File: python/algo/leecode2/algo282.py
class Solution:

    def addOperators(self, num: str, target: int):
        '''
        Use last multiple value instead of cacluate by precedence        
        '''
        n = len(num)
        
        ans = []
        # bfs array
        ops = [(0, "", 0, 0)]
        count = 0
        while ops:
            count += 1
            '''
            !!!!SOLVE THE OVERTIME!!!!!!!!!!!!!
            '''
            # inx, expr, ret, mul = ops.pop(0)
            inx, expr, ret, mul = ops.pop(-1)
            if inx == n:
                if ret == target:
                    ans.append(expr)
                continue
            
            for i in range(1, n+1-inx):
                if num[inx] == "0" and i > 1:break
                val = int(num[inx:i + inx])
                if inx == 0:
                    ops.append((i, num[inx:i + inx], val, val))
                else:
                    ops.append((i + inx, expr+"*"+num[inx:i + inx], ret - mul + mul * val, mul*val))
                    ops.append((i + inx, expr+"+"+num[inx:i + inx], ret+val, val))
                    ops.append((i + inx, expr+"-"+num[inx:i + inx], ret-val, -val))
        print(count)
        return ans


    def addOperators_recursion(self, num: str, target: int):
        '''
        The final solution:
        First, use recusion instead of bfs array, why?????????
        Secord, use last multiple value instead of cacluate by precedence
        '''
        n = len(num)
        
        ans = []
        count = 0
        def backtrack(inx, expr, ret, mul):
            nonlocal count
            count += 1
            if inx == n:
                if ret == target:
                    ans.append(expr)
                return
            
            for i in range(1, n+1-inx):
                if num[inx] == "0" and i > 1:break
                val = int(num[inx:i + inx])
                if inx == 0:
                    backtrack(i, num[inx:i + inx], val, val)
                else:
                    backtrack(i + inx, expr+"*"+num[inx:i + inx], ret - mul + mul * val, mul*val)
                    backtrack(i + inx, expr+"+"+num[inx:i + inx], ret+val, val)
                    backtrack(i + inx, expr+"-"+num[inx:i + inx], ret-val, -val)
        
        backtrack(0, "", 0, 0)
        print(count)
        return ans    

    def addOperators_original(self, num: str, target: int):
        '''
        !!!!!!!!!!!!!OVERTIME!!!!!!!!!!!!!
        '''
        multi = "*"
        add = "+"
        sub = "-"

        def calculate(stack) -> int:
            opd2, op, opd1 = int(stack.pop(-1)), stack.pop(-1), int(stack.pop(-1))

            if op == multi:
                stack.append(opd1 * opd2)
            if op == add:
                stack.append(opd1 + opd2)
            if op == sub:
                stack.append(opd1 - opd2)

        # If return True, pop op1 from prev to calcuate
        # If return False, push op2 into stack
        def precedence(op1:str, op2:str) -> bool:
            if op1 == multi:
                return True
            if op1 in [add, sub] and op2 in [add,sub]:
                return True
            if op1 in [add, sub] and op2 in [multi]:
                return False

        n = len(num)
        
        ret = []
        ops = [(0, "", [])]
        while ops:
            '''
            ops.pop(0) CHANGE TO ops.pop(-1)
            !!!!SOLVE THE OVERTIME!!!!!!!!!!!!!
            '''
            inx, expr, stack = ops.pop(-1)
            if inx == n:
                while len(stack) > 1:
                    calculate(stack)
                if stack[0] == target:
                    ret.append(expr)
                continue
            
            for i in range(1, n+1-inx):
                if num[inx] == "0" and i > 1:break
                if i == n - inx:
                    t_stack = stack.copy()
                    t_stack.append(int(num[inx:]))
                    ops.append((i + inx, expr+num[inx:i + inx], t_stack))
                    continue
                for op in [multi, add, sub]:
                    t_stack = stack.copy()
                    if t_stack and  precedence(t_stack[-1], op):
                        t_stack.append(int(num[inx:i+inx]))
                        calculate(t_stack)
                        while len(t_stack) > 1 and precedence(t_stack[-2], op):
                            calculate(t_stack)
                        t_stack.append(op)
                    else:
                        t_stack.append(num[inx:i+inx])
                        t_stack.append(op)
                    ops.append((i + inx, expr+num[inx:i + inx]+op, t_stack))
        
        return ret


    def addOperators_original_recursion(self, num: str, target: int):
        '''
        !!!!!!!!!!!!!RECURSION IS MUST!!!!!!!!!!!!
        '''        
        multi = "*"
        add = "+"
        sub = "-"

        def calculate(stack) -> int:
            opd2, op, opd1 = int(stack.pop(-1)), stack.pop(-1), int(stack.pop(-1))

            if op == multi:
                stack.append(opd1 * opd2)
            if op == add:
                stack.append(opd1 + opd2)
            if op == sub:
                stack.append(opd1 - opd2)

        # If return True, pop op1 from prev to calcuate
        # If return False, push op2 into stack
        def precedence(op1:str, op2:str) -> bool:
            if op1 == multi:
                return True
            if op1 in [add, sub] and op2 in [add, sub]:
                return True
            if op1 in [add, sub] and op2 in [multi]:
                return False

        n = len(num)
        
        ret = []
        def backtrack(inx, expr, stack):

            if inx == n:
                while len(stack) > 1:
                    calculate(stack)
                if stack[0] == target:
                    ret.append(expr)
                return
            
            for i in range(1, n+1-inx):
                if num[inx] == "0" and i > 1:break
                if i == n - inx:
                    t_stack = stack.copy()
                    t_stack.append(int(num[inx:]))
                    backtrack(i + inx, expr+num[inx:i + inx], t_stack)
                    continue
                for op in [multi, add, sub]:
                    t_stack = stack.copy()
                    if t_stack and  precedence(t_stack[-1], op):
                        t_stack.append(int(num[inx:i+inx]))
                        calculate(t_stack)
                        while len(t_stack) > 1 and precedence(t_stack[-2], op):
                            calculate(t_stack)                        
                        t_stack.append(op)
                    else:
                        t_stack.append(int(num[inx:i+inx]))
                        t_stack.append(op)
                    backtrack(i + inx, expr+num[inx:i + inx]+op, t_stack)
        
        backtrack(0, "", [])
        return ret

    def addOperators_eval(self, num: str, target: int):
        '''
        !!!!!!!!!!!!!OVERTIME!!!!!!!!!!!!!
        !!!PERFROMANCE OF EVAL IS WORST!!!
        '''
        n = len(num)
        multi, add, sub = "*", "+", "-"

        ret = []
        ops = [(0, "")]
        while ops:
            inx, expr = ops.pop(0)
            if inx == n:
                if eval(expr) == target:
                    ret.append(expr)
                continue
            
            for i in range(1, n+1-inx):
                if num[inx] == "0" and i > 1:break
                if i == n - inx:
                    ops.append((i + inx, expr+num[inx:i + inx]))
                    continue
                for op in [multi, add, sub]:
                    ops.append((i + inx, expr+num[inx:i + inx]+op))
        
        return ret

    def addOperators_eval_recursion(self, num: str, target: int):
        '''
        !!SHOULD PASS WITH BAD PERFORMANCE!!
        '''
        n = len(num)
        multi, add, sub = "*", "+", "-"

        ret = []
        def backtrack(inx, expr):
            if inx == n:
                if eval(expr) == target:
                    ret.append(expr)
                return
            
            for i in range(1, n+1-inx):
                if num[inx] == "0" and i > 1:break
                if i == n - inx:
                    backtrack(i + inx, expr+num[inx:i + inx])
                    continue
                for op in [multi, add, sub]:
                    backtrack(i + inx, expr+num[inx:i + inx]+op)
        backtrack(0, "")
        return ret

if __name__ == "__main__":
    sol = Solution()
    num, target = "123", 6
    num, target = "1054", 5
    num, target = "3456237490", 9191
    num, target = "123456789", 45
    from datetime import datetime
    # begin = datetime.now()
    # print(sol.addOperators(num, target))
    # print((datetime.now()- begin).total_seconds())

    # begin = datetime.now()
    # print(sol.addOperators_recursion(num, target))
    # print((datetime.now()- begin).total_seconds())

    # begin = datetime.now()
    # print(sol.addOperators_original(num, target))
    # print((datetime.now()- begin).total_seconds())

    begin = datetime.now()
    print(sol.addOperators_original_recursion(num, target))
    print((datetime.now()- begin).total_seconds())

    begin = datetime.now()
    print(sol.addOperators_original(num, target))
    print((datetime.now()- begin).total_seconds())

File: cplus/leetcode/algo282.cpp
#include <iostream>
#include <vector>
#include <functional>

using namespace std;

struct element {
    int inx;
    string expr;
    long res;
    long mul;
};

class Solution {
public:
    vector<string> addOperators(string num, int target) {
        int n = num.length();
        vector<string> ans;
        
        vector<element> ops;

        ops.push_back(element{0, "", 0, 0});
        while (!ops.empty()) {
            // change the ops put/pop from queue to stack
            // element cur = ops.front();
            // ops.erase(ops.begin());
            element cur = ops.back();
            ops.erase(ops.end());
            if (cur.inx == n) {
                if (cur.res == target) {
                    ans.push_back(cur.expr);
                }
                continue;
            }
            for (int i = 1; i <= n - cur.inx; i++) {
                if (num[cur.inx] == '0' && i > 1) {
                    break;
                }
                long val = stol(num.substr(cur.inx, i));
                if (cur.inx == 0) {
                    ops.push_back(element{i, num.substr(cur.inx, i), val, val});
                } else {
                    ops.push_back(element{i + cur.inx, cur.expr + "*" + num.substr(cur.inx, i), cur.res - cur.mul + cur.mul * val, cur.mul * val});
                    ops.push_back(element{i + cur.inx, cur.expr + "+" + num.substr(cur.inx, i), cur.res + val, val});
                    ops.push_back(element{i + cur.inx, cur.expr + "-" + num.substr(cur.inx, i), cur.res - val, -val});
                }
            }
        }
        return ans;
    }

    vector<string> addOperators_recursion(string num, int target) {
        int n = num.length();
        vector<string> ans;
        
        function<void(string, int, long, long)> backtrack = [&](string expr, int inx, long res, long mul) {
            if (inx == n) {
                if (res == target) {
                    ans.push_back(expr);
                }
                return;
            }
            for (int i = 1; i <= n - inx; i++) {
                if (num[inx] == '0' && i > 1) {
                    break;
                }
                long val = stol(num.substr(inx, i));
                if (inx == 0) {
                    backtrack(num.substr(inx, i), i, val, val);
                } else {
                    backtrack(expr + "*" + num.substr(inx, i), i + inx, res - mul + mul * val, mul * val);
                    backtrack(expr + "+" + num.substr(inx, i), i + inx, res + val, val);
                    backtrack(expr + "-" + num.substr(inx, i), i + inx, res - val, -val);
                }
            }
        };
        string expr;
        backtrack(expr, 0, 0, 0);
        return ans;
    }

    /// expr transfer string to string&
    vector<string> addOperators3(string num, int target) {
        int n = num.length();
        vector<string> ans;
        
        function<void(string&, int, long, long)> backtrack = [&](string& expr, int inx, long res, long mul) {
            if (inx == n) {
                if (res == target) {
                    ans.push_back(expr);
                }
                return;
            }
            int signIndex = expr.size();
            if (inx > 0) {
                expr.push_back(0);
            }
            for (int i = 1; i <= n - inx; i++) {
                if (num[inx] == '0' && i > 1) {
                    break;
                }
                long val = stol(num.substr(inx, i));
                expr.push_back(num[inx+i-1]);
                if (inx == 0) {
                    backtrack(expr, i, val, val);
                } else {
                    expr[signIndex] = '*'; backtrack(expr, i + inx, res - mul + mul * val, mul * val);
                    expr[signIndex] = '+'; backtrack(expr, i + inx, res + val, val);
                    expr[signIndex] = '-'; backtrack(expr, i + inx, res - val, -val);
                }
            }
            expr.resize(signIndex);
        };
        string expr;
        backtrack(expr, 0, 0, 0);
        return ans;
    }
};



int main() {
    Solution solution;
    vector<string> result = solution.addOperators("123456789", 45);
    for (const string& expr : result) {
        cout << expr << endl;
    }
    return 0;
}